### PR TITLE
Special-case unreachable If conditions in IRBuilder

### PR DESCRIPTION
**Pull request by tlively**
When IRBuilder completes a scope that contains an unreachable
expression, it has to find the last unreachable expression and drop any
concretely typed expression that precedes it. If the original
unreachable expression has been made a child of some other expression,
that's usually fine because that other expression will become
unreachable as well. However, if conditions are special because they can
be unreachable without making their ifs unreachable, so we have to work
harder to find unreachable if conditions.

### CODE REVIEW

**Comment by tlively** [comment]
This is a simpler but arguably hackier alternative to #7094.

File: src/wasm/wasm-ir-builder.cpp [hunk 1]
```diff
@@ -759,13 +759,23 @@ Result<Expression*> IRBuilder::finishScope(Block* block) {
...
+      } else {
+        // As a special case, look for unreachable if conditions, since they may
+        // not make their ifs unreachable.
+        for (auto* curr = scope.exprStack[i]->dynCast<If>(); curr;
+             curr = curr->condition->dynCast<If>()) {
+          if (curr->condition->type == Type::unreachable) {
+            sawUnreachable = true;
+            break;
+          }
+        }
```
Comment (kripken) [inline src/wasm/wasm-ir-builder.cpp:778]: Would it be simpler to change this?

https://github.com/WebAssembly/binaryen/blob/e13bf0fb72fca160f457570b930c4ba3c35ead3a/src/wasm/wasm.cpp#L230

I think we could just make an if with an unreachable condition be unreachable, unless I'm missing a reason not to.

Comment (kripken) [reply src/wasm/wasm-ir-builder.cpp:778]: (that would avoid special-casing here, and seems simpler overall: the code is dead, so let's mark it dead)

**Comment by tlively** [comment]
Closing this in favor of #7094.

**Pull request closed without merging**